stream/replicator.h

```diff
@@ -80,36 +82,155 @@ class SubscribableRemoteStream final {
+             current::ToString(current::time::Now().count()) + (mode == SubscriptionMode::Checked ? "&checked" : "");
+    }
+
    private:
     const std::string url_;
     const SubscribableStreamSchema schema_;
   };
 
-  template <typename F, typename TYPE_SUBSCRIBED_TO, SubscriptionMode SM>
-  class RemoteSubscriberThread final : public current::stream::SubscriberScope::SubscriberThread {
+  template <typename F, typename TYPE_SUBSCRIBED_TO, ReplicationMode RM>
+  class RemoteStreamSubscriber {
     static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
 
+   public:
+    RemoteStreamSubscriber(Borrowed<RemoteStream> remote_stream,
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::chrono::microseconds from_us = std::chrono::microseconds(0),
+                           std::function<void()> destruction_callback = nullptr)
+        : borrowed_remote_stream_(std::move(remote_stream), destruction_callback),
+          subscriber_(subscriber),
+          index_(start_idx),
+          from_us_(from_us),
+          unused_idxts_() {}
+
+    void PassChunkToSubscriber(const std::string& chunk) {
+      const size_t chunk_size = chunk.size();
+      size_t end_pos = 0;
+      if (!carried_over_data_.empty()) {
+        while (end_pos < chunk_size && chunk[end_pos] != '\n' && chunk[end_pos] != '\r') {
+          ++end_pos;
+        }
+        if (end_pos == chunk_size) {
+          carried_over_data_ += chunk;
+          return;
+        }
+        PassEntryToSubscriber(carried_over_data_ + chunk.substr(0, end_pos));
+      }
+
+      size_t start_pos = end_pos;
+      for (;;) {
+        while (start_pos < chunk_size && (chunk[start_pos] == '\n' || chunk[start_pos] == '\r')) {
+          ++start_pos;
+        }
+        end_pos = start_pos + 1;
+        while (end_pos < chunk_size && chunk[end_pos] != '\n' && chunk[end_pos] != '\r') {
+          ++end_pos;
+        }
+        if (end_pos >= chunk_size) {
+          break;
+        }
+        PassEntryToSubscriber(chunk.substr(start_pos, end_pos - start_pos));
+        start_pos = end_pos + 1;
+      }
+      if (start_pos < chunk_size) {
+        carried_over_data_ = chunk.substr(start_pos);
+      } else {
+        carried_over_data_.clear();
+      }
+    }
+
+   protected:
+    BorrowedWithCallback<RemoteStream> borrowed_remote_stream_;
+    F& subscriber_;
+    uint64_t index_;
+    std::chrono::microseconds from_us_;
+    const idxts_t unused_idxts_;
+    std::string carried_over_data_;
+
+   private:
+    template <ReplicationMode MODE = RM>
+    ENABLE_IF<MODE == ReplicationMode::Checked> PassEntryToSubscriber(const std::string& raw_log_line) {
+      const auto split = current::strings::Split(raw_log_line, '\t');
+      if (split.empty()) {
+        CURRENT_THROW(RemoteStreamMalformedChunkException());
+      }
+      const auto tsoptidx = ParseJSON<ts_optidx_t>(split[0]);
```

Comment:
I understand it's copy-pasted (well, moved!), but, in `Checked` mode, shoudn't we also proxy a possible JSON parsing exception as an `RemoteStreamMalformedChunkException` one?